src/Npgsql/NpgsqlConnector.Auth.cs

```diff
@@ -362,32 +362,35 @@ class AuthenticationCompleteException : Exception { }
 
         string? GetPassword(string username)
         {
-            var passwd = Settings.Password;
-            if (passwd != null)
-                return passwd;
+            var password = Settings.Password;
+            if (password != null)
+                return password;
 
             // No password was provided. Attempt to pull the password from the pgpass file.
-            var matchingEntry = PgPassFile.Load(Settings.Passfile)?.GetFirstMatchingEntry(Host, Port, Settings.Database!, username);
-            if (matchingEntry != null)
+            var passFilePath = Settings.Passfile ?? PostgresEnvironment.PassFile ?? PostgresEnvironment.PassFileDefault;
+            if (passFilePath != null && File.Exists(passFilePath))
             {
-                Log.Trace("Taking password from pgpass file");
-                return matchingEntry.Password;
-            }
-
-            if (ProvidePasswordCallback != null)
-            {
-                Log.Trace($"Taking password from {nameof(ProvidePasswordCallback)} delegate");
-                try
-                {
-                    return ProvidePasswordCallback(Host, Port, Settings.Database!, username);
-                }
-                catch (Exception e)
+                var matchingEntry = new PgPassFile(passFilePath)
+                    .GetFirstMatchingEntry(Host, Port, Settings.Database!, username);
```

Comment:
Should database be passed with a bang here? Or should we explicitly coalesce it to the username we've resolved?

Reply:
`Database` property is nullable, so yes, it should be passed with bang. But it doesn't mean that we should coalesce here with the username, because it could already contain it.

https://github.com/npgsql/npgsql/blob/4aac0e5ef265e049b390babceae7493debfa8b51/src/Npgsql/NpgsqlConnector.cs#L374-L376

Reply:
Ah, of course. Thanks for clarifying!